### ISSUE TITLE
Data URIs with more than 65519 characters are always stripped

### ISSUE THREAD

**Issue by chieftain93** [mganss/HtmlSanitizer#66]
I'm trying to embed images (base64 encoded) in some html that's stored in a db. The encoded data is stored in the src attribute. The src attribute is supposed to be included by default but when I run the sanitizer it removes the src and its contents. The attribute looks something like this: ``src="data:image/png;base64,iVBORw0KGgoAAAANSU

Is there a way to do this? I even tried adding the src attribute to my sanitizer. Thanks.

**Comment by mganss** [comment]
You need to add the data scheme to the set of allowed protocol schemes:

```C#
var s = new HtmlSanitizer();
s.AllowedSchemes.Add("data");
```

**Comment by chieftain93** [comment]
Thanks. That works for some embedded images but not all. Of my six test cases about half have their data stripped. It seems that these images are larger than 42KB. Is there an upper size limit imposed or is it possible that these larger images have content that triggers the removal?

**Comment by mganss** [comment]
Can you share an example?

**Comment by chieftain93** [comment]
Here's an example. The file contains the img tag and its attributes including the data.
[image base 64 encoded.txt](https://github.com/mganss/HtmlSanitizer/files/206178/image.base.64.encoded.txt)

**Comment by mganss** [comment]
The example works for me, i.e. output is the same as input. The data URL in the input seems to be broken, though. Chrome shows the broken image icon and the error message "net::ERR_INVALID_URL" in the console.

Do you have the latest version of HtmlSanitizer and AngleSharp?

**Comment by iernie** [comment]
I am having the same issue. If the base64 encoding is really big, the src attribute gets omitted from the sanitized html. Works for small pictures.

**Comment by mganss** [comment]
@iernie Can you provide an example that shows the issue using the latest version?

**Comment by iernie** [comment]
Sure!

This is my setup
```
var sanitizer = new HtmlSanitizer();
sanitizer.AllowedSchemes.Add("data");
sanitizer.AllowedSchemes.Add("mailto");
sanitizer.AllowedAttributes.Add("class");
sanitizer.AllowDataAttributes = true;
```

Here are some examples. Ignore the data-mce-src, it's there because this comes from a tinymce editor.

#### Example that doesn't work

This image is 48,8 KB in size, jpg.

Getting this **[input](http://pastebin.com/mnhemFSX)**
Produces this **[result](http://pastebin.com/1qBQV5wc)**


#### Example that does work

It seems to work with some images, seems like smaller ones. This image is 38,1 KB in size, jpg.

Getting this **[input](http://pastebin.com/srvv8tE4)**
Produces this **[result](http://pastebin.com/P1xRP3q5)**

**Comment by mganss** [comment]
This occurs because when we're sanitizing the URI, we're trying to create a `System.Uri` object (in [`GetSafeUri()`](https://github.com/mganss/HtmlSanitizer/blob/d33cb8672ba0969c1e0b8e75c0db9f8ed9bf5cee/HtmlSanitizer/HtmlSanitizer.cs#L681-L695)) and this fails because `System.Uri` has a [limit of 65519 characters](https://msdn.microsoft.com/en-us/library/z6c2z492.aspx):

> The length of uriString exceeds 65519 characters.

**Comment by iernie** [comment]
Aha, any way around it?

**Comment by mganss** [comment]
This:

```C#
sanitizer.RemovingAttribute += (s, e) => e.Cancel = e.Reason == RemoveReason.NotAllowedUrlValue
    && e.Attribute.Value.Length >= 0xfff0 
    && e.Attribute.Value.StartsWith("data:", StringComparison.OrdinalIgnoreCase);
```

Note that this might also allow invalid URIs which carry XSS potential.

**Comment by iernie** [comment]
Thank you, that worked. In my case this is an internal system, so XSS is less likely, but I understand the reasons for doing it.

**Comment by tiesont** [comment]
Also worth noting (if you're required to support it) that IE (and Edge) have some limits to the size of the data URIs that they can render: http://caniuse.com/#feat=datauri

**Comment by mganss** [comment]
Interesting, though 4GB in Edge should be sufficient for most use cases 😄 I wonder if someone has really tested the other browsers which supposedly have full support to work with more than 4GB, e.g. Opera Mini...

**Comment by hardhub** [comment]
>Also worth noting (if you're required to support it) that IE (and Edge) have some limits to the size of data URIs that they can render: http://caniuse.com/#feat=datauri

Limit 4GB :)

>Note that this might also allow invalid URIs which carry XSS potential.

Maybe it is possible to implement by different way?   
Maybe Regex-based GetSafeUri? Or custom parsing? 3rd party?

**Comment by mganss** [comment]
It's been fixed already in 255fd58 😸

**Comment by hardhub** [comment]
It does not work by default.. without sanitizer.RemovingAttribute trick.

**Comment by mganss** [comment]
Which version are you using? The fix is available from 3.5.168-beta.

**Comment by hardhub** [comment]
latest stable 3.4.156

**Comment by alienown** [comment]
Am I right that when declaring font-face with src that is base64 it still doesn't work? I'm trying to get it to work but for large base64 after sanitizing the src is empty.

The following example works:
```
@font-face {
    font-family: "MyFont";
    src: url("data:font/woff2;charset=utf-8;base64,d09GMgABAAAAAJvEABQAAAABTXAAAJtYAAAAAAAAAAAAAA") format('woff');
    font-weight: normal;
    font-style: normal;
    font-display: swap;
}
```

But when you insert src that is large enough it gets cleared. Is there any workaround?

**Comment by mganss** [comment]
@alienown This seems to be an issue with AngleSharp.Css (see https://github.com/AngleSharp/AngleSharp.Css/issues/76). In my tests, the limit seems to be 24293 characters 🤷🏻 

BTW, as stated in the README, `@font-face` declarations are not sanitized by HtmlSanitizer: https://github.com/mganss/HtmlSanitizer#css-at-rules-allowed-by-default

**Comment by alienown** [comment]
@mganss I must have missed that paragraph in README 😩. Anyway, thanks for really quick reply! I'll track the issue you have created 🙇‍♂️.

**Comment by mganss** [comment]
@alienown This should be fixed in 6.0.441 via AngleSharp 0.16.1.
